--- ./emacs/merlin.el
+++ ./emacs/merlin.el
@@ -1651,7 +1651,7 @@
     (when (equal command 'opam)
       (with-temp-buffer
         (if (eq (call-process-shell-command
-                 "opam config var bin" nil (current-buffer) nil) 0)
+                 (if (eq system-type 'cygwin) "cygpath \"$(opam config var bin)\"" "opam config var bin") nil (current-buffer) nil) 0)
             (progn
               (setq command (concat
                              (replace-regexp-in-string "\n$" "" (buffer-string))
--- ./src/frontend/ocamlmerlin.c
+++ ./src/frontend/ocamlmerlin.c
@@ -537,7 +537,7 @@
     user[0] = '\0';
   // @@DRA Need to use Windows API functions to get meaningful values for st_dev and st_ino
   snprintf(eventname, PATHSZ,
-      "ocamlmerlin_%s_%lx_%llx",
+      "ocamlmerlin_%s_%lx_%I64x",
       user,
       info.dwVolumeSerialNumber,
       ((__int64)info.nFileIndexHigh) << 32 | ((__int64)info.nFileIndexLow));
